src/cliffordsynthesis/CliffordSynthesizer.cpp

```diff
@@ -0,0 +1,331 @@
+            } else {
+                HeuristicStrategy::runHeuristicStrategy(subset, qubitSet, configuration, *this);
+            }
+
+            if (configuration.chooseBest && optimalResults.sat) {
+                break;
+            }
+        }
+
+        auto                          totalEnd = std::chrono::high_resolution_clock::now();
+        std::chrono::duration<double> diff     = totalEnd - totalStart;
+        INFO() << "Time: " << diff.count() << std::endl;
+        optimalResults.totalSeconds = diff.count();
+    }
+
+    Results CliffordSynthesizer::mainOptimization(
+            std::size_t          timesteps,
+            const CouplingMap&   reducedCM,
+            const QubitSubset&   qubitChoice,
+            const Tableau&       targetTableau,
+            const Tableau&       initialTableau,
+            const Configuration& configuration) {
+        using namespace logicbase;
+        std::unique_ptr<LogicBlock> lb;
+        bool                        success = false;
+        if (configuration.method == ReasoningEngine::Z3) {
+            LogicTerm::termType = TermType::BASE;
+            if (configuration.strategy == OptimizationStrategy::UseMinimizer || configuration.strategy == OptimizationStrategy::SplitIter) {
+                logicutil::Params params;
+                params.addParam("pb.compile_equality", true);
+                params.addParam("maxres.hill_climb", true);
+                params.addParam("maxres.pivot_on_correction_set", false);
+                lb = logicutil::getZ3LogicOptimizer(success, true, params);
+            } else {
+                logicutil::Params params;
+                params.addParam("threads", static_cast<unsigned>(configuration.nThreads / 2));
+                lb = logicutil::getZ3LogicBlock(success, true, params);
+            }
+        } else {
+            return Results{};
+        }
+        if (!success) {
+            throw QMAPException("Could not initialize Z3 logic block optimizer");
+        }
+        DEBUG() << "lb 1: " << lb.get() << std::endl;
+        logicbase::LogicMatrix   x{};
+        logicbase::LogicMatrix   z{};
+        logicbase::LogicVector   r{};
+        logicbase::LogicMatrix3D gS{};
+        logicbase::LogicMatrix3D gC{};
+
+        auto changes = logicbase::LogicTerm(true);
+
+        auto start = std::chrono::high_resolution_clock::now();
+        /*
+   * Tableau Variables x/z
+   * k before gate k is applied
+   * i column
+   */
+        std::stringstream xName{};
+        std::stringstream zName{};
+        std::stringstream rName{};
+        for (std::size_t k = 0U; k < timesteps + 1U; ++k) {
+            x.emplace_back();
+            z.emplace_back();
+            for (std::size_t i = 0U; i < configuration.nqubits; ++i) {
+                xName.str("");
+                zName.str("");
+                xName << "x_" << k << "_" << i;
+                zName << "z_" << k << "_" << i;
+                x.back().push_back(
```

Comment:
why not use `emplace_back` here as well? same question holds for all similar calls down below.